```bash
f9ab3fa4c Merge 8c779d8b7923b080d58dc7ca1f856afb6249bbb7 into 9c8790667060ec78056ee532138380fc78bb807c
8c779d8b7 fix(image): image scale add expand pixel
Remove Gerrit Change-ID's before submitting upstream
Error: Process completed with exit code 1.
```

--- a/lv_conf_template.h
+++ b/lv_conf_template.h
@@ -768,6 +768,9 @@
  *Requires `LV_USE_MATRIX = 1`*/
 #define LV_USE_VECTOR_GRAPHIC  0
 
+/*Enable the image decoder size expand feature*/
+#define LV_DECODER_IMG_SIZE_EXPAND 0
+
 /* Enable ThorVG (vector graphics library) from the src/libs folder */
 #define LV_USE_THORVG_INTERNAL 0
 

--- a/src/draw/lv_draw_buf.c
+++ b/src/draw/lv_draw_buf.c
@@ -542,6 +542,122 @@ void lv_draw_buf_set_palette(lv_draw_buf_t * draw_buf, uint8_t index, lv_color32
     lv_memcpy(&buf[index * sizeof(color)], &color, sizeof(color));
 }
 
+void lv_draw_buf_copy_palette(lv_draw_buf_t * dest, lv_draw_buf_t * src)
+{
+    if(dest->header.cf != src->header.cf) {
+        return;
+    }
+
+    lv_color_format_t cf = src->header.cf;
+    uint32_t size = LV_COLOR_INDEXED_PALETTE_SIZE(cf);
+    if(size == 0) {
+        return;
+    }
+
+    lv_memcpy(dest->data, src->data, size * sizeof(lv_color32_t));
+}
+
+lv_draw_buf_t * lv_draw_buf_fill_expand(lv_draw_buf_t * draw_buff, uint32_t expand_size)
+{
+    lv_draw_buf_t * new_draw_buf = NULL;
+    lv_color_format_t cf = draw_buff->header.cf;
+    uint32_t expand_width = draw_buff->header.w + expand_size * 2;
+    uint32_t expand_height = draw_buff->header.h + expand_size * 2;
+
+    if(!LV_COLOR_FORMAT_IS_INDEXED(cf)) {
+        new_draw_buf = lv_draw_buf_create(expand_width, expand_height, cf, LV_STRIDE_AUTO);
+        lv_draw_buf_clear(new_draw_buf, NULL);
+    }
+    else if(cf == LV_COLOR_FORMAT_I8) {
+        uint32_t palette_size = LV_COLOR_INDEXED_PALETTE_SIZE(cf);
+        int color_index = -1;
+        lv_color32_t * data = (lv_color32_t *)draw_buff->data;
+        for(int i = 0; i < palette_size; i++) {
+            if(data->alpha == 0) {
+                color_index = i;
+                break;
+            }
+            data++;
+        }
+
+        if(color_index < 0) {
+            LV_LOG_WARN("image has no transparent color in palette!, use RGBA8888");
+            new_draw_buf = lv_draw_buf_create(expand_width, expand_height, LV_COLOR_FORMAT_ARGB8888, LV_STRIDE_AUTO);
+            lv_draw_buf_clear(new_draw_buf, NULL);
+        }
+        else {
+            new_draw_buf = lv_draw_buf_create(expand_width, expand_height, cf, LV_STRIDE_AUTO);
+
+            /*clear buffer*/
+            uint8_t * dest = lv_draw_buf_goto_xy(new_draw_buf, 0, 0);
+            lv_memset(dest, color_index, new_draw_buf->header.stride * new_draw_buf->header.h);
+        }
+    }
+
+    if(new_draw_buf) {
+        lv_draw_buf_copy_palette(new_draw_buf, draw_buff);
+    }
+
+    return new_draw_buf;
+}
+
+void lv_draw_buf_color_convert(lv_draw_buf_t * dst, lv_draw_buf_t * src, uint32_t dst_offset_x, uint32_t dst_offset_y)
+{
+    uint8_t * src_buf = lv_draw_buf_goto_xy(src, 0, 0);
+    uint8_t * dst_buf = lv_draw_buf_goto_xy(dst, 0, 0);
+    uint32_t  dst_pixel_size = lv_color_format_get_size(dst->header.cf);
+    uint32_t  src_pixel_size = lv_color_format_get_size(src->header.cf);
+    int x;
+    int y;
+
+    dst_buf += dst_offset_y * dst->header.stride + dst_offset_x * dst_pixel_size;
+    if(dst->header.cf == src->header.cf) {
+        for(y = 0; y < src->header.h; y++) {
+            lv_memcpy(dst_buf, src_buf, src->header.w * src_pixel_size);
+            dst_buf += dst->header.stride;
+            src_buf += src->header.stride;
+        }
+    }
+    else {
+        if(src->header.cf == LV_COLOR_FORMAT_I8 && dst->header.cf == LV_COLOR_FORMAT_ARGB8888) {
+            lv_color32_t * paltte_head = (lv_color32_t *)src->data;
+            for(y = 0; y < src->header.h; y++) {
+                lv_color32_t * dst_pixel = (lv_color32_t *)dst_buf;
+                const uint8_t * src_pixel = src_buf;
+                for(x = 0; x < src->header.w; x++) {
+                    *dst_pixel = paltte_head[*src_pixel];
+                    dst_pixel++;
+                    src_pixel++;
+                }
+                dst_buf += dst->header.stride;
+                src_buf += src->header.stride;
+            }
+        }
+        else {
+            LV_LOG_WARN("color convert not implemented yet");
+        }
+    }
+}
+
+lv_draw_buf_t * lv_draw_buf_expand(lv_draw_buf_t * decoded, uint32_t expand_size)
+{
+    if(expand_size <= 0) {
+        return decoded;
+    }
+
+    lv_draw_buf_t * new_decoded = lv_draw_buf_fill_expand(decoded, expand_size);
+
+    if(!new_decoded) {
+        LV_LOG_WARN("expand for decoded image failed!");
+        return decoded;
+    }
+
+    lv_draw_buf_color_convert(new_decoded, decoded, expand_size, expand_size);
+    lv_draw_buf_destroy(decoded);
+
+    return new_decoded;
+}
+
 /**********************
  *   STATIC FUNCTIONS
  **********************/

--- a/src/draw/lv_draw_buf.h
+++ b/src/draw/lv_draw_buf.h
@@ -330,6 +330,31 @@ lv_result_t lv_draw_buf_adjust_stride(lv_draw_buf_t * src, uint32_t stride);
  */
 lv_result_t lv_draw_buf_premultiply(lv_draw_buf_t * draw_buf);
 
+/**
+ * Copy the palette from `src` to `dest` draw buffer.
+ */
+void lv_draw_buf_copy_palette(lv_draw_buf_t * dest, lv_draw_buf_t * src);
+
+/**
+ * Convert the color format of the draw buffer to dst draw buffer.
+ */
+void lv_draw_buf_color_convert(lv_draw_buf_t * dst, lv_draw_buf_t * src, uint32_t dst_offset_x, uint32_t dst_offset_y);
+
+/**
+ * Alloc and Fill the draw buffer with the alpha color.
+ */
+lv_draw_buf_t * lv_draw_buf_fill_expand(lv_draw_buf_t * draw_buff, uint32_t expand_size);
+
+/**
+ * Expand the draw buffer by the given size.
+ * The new area will be filled with the alpha color. this function workaroud the issue of GPU drawing
+ * @param origin       the original draw buffer
+ * @param expand_size  the size to expand in pixels
+ *
+ * @return the expanded draw buffer on success, origin buffer will be return if failed.
+ */
+lv_draw_buf_t * lv_draw_buf_expand(lv_draw_buf_t * origin, uint32_t expand_size);
+
 static inline bool lv_draw_buf_has_flag(const lv_draw_buf_t * draw_buf, lv_image_flags_t flag)
 {
     return draw_buf->header.flags & flag;

--- a/src/draw/lv_image_decoder.c
+++ b/src/draw/lv_image_decoder.c
@@ -288,6 +288,16 @@ lv_draw_buf_t * lv_image_decoder_post_process(lv_image_decoder_dsc_t * dsc, lv_d
     return decoded;
 }
 
+void lv_image_decoder_header_expand(lv_image_header_t * header, uint32_t expand_size)
+{
+    header->w += expand_size * 2;
+    header->h += expand_size * 2;
+
+    if(expand_size > 0) {
+        header->flags |= LV_IMAGE_FLAGS_EXPANDED;
+    }
+}
+
 /**********************
  *   STATIC FUNCTIONS
  **********************/

--- a/src/draw/lv_image_decoder.h
+++ b/src/draw/lv_image_decoder.h
@@ -296,6 +296,13 @@ lv_cache_entry_t * lv_image_decoder_add_to_cache(lv_image_decoder_t * decoder,
  */
 lv_draw_buf_t * lv_image_decoder_post_process(lv_image_decoder_dsc_t * dsc, lv_draw_buf_t * decoded);
 
+/**
+ * Expand the image header by the given size.
+ * @param header the image info will be stored here
+ * @param expand_size the size to expand
+ */
+void lv_image_decoder_header_expand(lv_image_header_t * header, uint32_t expand_size);
+
 /**********************
  *      MACROS
  **********************/

--- a/src/draw/lv_image_dsc.h
+++ b/src/draw/lv_image_dsc.h
@@ -56,6 +56,11 @@ typedef enum _lv_image_flags_t {
     LV_IMAGE_FLAGS_MODIFIABLE       = 0x0020,
 
     /**
+     * The image has expanded pixel data.
+     */
+    LV_IMAGE_FLAGS_EXPANDED          = 0x0040,
+
+    /**
      * Flags reserved for user, lvgl won't use these bits.
      */
     LV_IMAGE_FLAGS_USER1            = 0x0100,

--- a/src/libs/libjpeg_turbo/lv_libjpeg_turbo.c
+++ b/src/libs/libjpeg_turbo/lv_libjpeg_turbo.c
@@ -23,7 +23,7 @@
 
 #define image_cache_draw_buf_handlers &(LV_GLOBAL_DEFAULT()->image_cache_draw_buf_handlers)
 
-#define JPEG_PIXEL_SIZE 4 /* XRGB888 */
+#define JPEG_PIXEL_SIZE 4 /* ARGB8888 */
 #define JPEG_SIGNATURE 0xFFD8FF
 #define IS_JPEG_SIGNATURE(x) (((x) & 0x00FFFFFF) == JPEG_SIGNATURE)
 
@@ -142,6 +142,8 @@ static lv_result_t decoder_info(lv_image_decoder_t * decoder, lv_image_decoder_d
         header->cf = LV_COLOR_FORMAT_RGB888;
         header->w = (orientation % 180) ? height : width;
         header->h = (orientation % 180) ? width : height;
+
+        lv_image_decoder_header_expand(header, LV_DECODER_IMG_SIZE_EXPAND);
 
         return LV_RESULT_OK;
     }
@@ -339,7 +341,7 @@ static lv_draw_buf_t * decode_jpeg_file(const char * filename)
 
     /* set parameters for decompression */
 
-    cinfo.out_color_space = JCS_EXT_BGRX;
+    cinfo.out_color_space = JCS_EXT_BGRA;
 
     /* In this example, we don't need to change any of the defaults set by
      * jpeg_read_header(), so we do nothing here.
@@ -413,6 +415,8 @@ static lv_draw_buf_t * decode_jpeg_file(const char * filename)
     /* At this point you may want to check to see whether any corrupt-data
     * warnings occurred (test whether jerr.pub.num_warnings is nonzero).
     */
+
+    decoded = lv_draw_buf_expand(decoded, LV_DECODER_IMG_SIZE_EXPAND);
 
     /* And we're done! */
     return decoded;

--- a/src/libs/libpng/lv_libpng.c
+++ b/src/libs/libpng/lv_libpng.c
@@ -107,6 +107,8 @@ static lv_result_t decoder_info(lv_image_decoder_t * decoder, lv_image_decoder_d
         header->w = (int32_t)((size[0] & 0xff000000) >> 24) + ((size[0] & 0x00ff0000) >> 8);
         header->h = (int32_t)((size[1] & 0xff000000) >> 24) + ((size[1] & 0x00ff0000) >> 8);
 
+        lv_image_decoder_header_expand(header, LV_DECODER_IMG_SIZE_EXPAND);
+
         return LV_RESULT_OK;
     }
 

--- a/src/libs/lodepng/lv_lodepng.c
+++ b/src/libs/lodepng/lv_lodepng.c
@@ -114,6 +114,8 @@ static lv_result_t decoder_info(lv_image_decoder_t * decoder, lv_image_decoder_d
             header->w = (int32_t)((size[0] & 0xff000000) >> 24) + ((size[0] & 0x00ff0000) >> 8);
             header->h = (int32_t)((size[1] & 0xff000000) >> 24) + ((size[1] & 0x00ff0000) >> 8);
 
+            lv_image_decoder_header_expand(header, LV_DECODER_IMG_SIZE_EXPAND);
+
             return LV_RESULT_OK;
         }
     }

--- a/src/lv_conf_internal.h
+++ b/src/lv_conf_internal.h
@@ -28,6 +28,8 @@
 #define LV_DRAW_SW_ASM_HELIUM       2
 #define LV_DRAW_SW_ASM_CUSTOM       255
 
+#define LV_ARRAY_DEFAULT_CAPACITY   8
+
 /* Handle special Kconfig options */
 #ifndef LV_KCONFIG_IGNORE
     #include "lv_conf_kconfig.h"
@@ -2505,6 +2507,15 @@
         #define LV_USE_VECTOR_GRAPHIC CONFIG_LV_USE_VECTOR_GRAPHIC
     #else
         #define LV_USE_VECTOR_GRAPHIC  0
+    #endif
+#endif
+
+/*Enable the image decoder size expand feature*/
+#ifndef LV_DECODER_IMG_SIZE_EXPAND
+    #ifdef CONFIG_LV_DECODER_IMG_SIZE_EXPAND
+        #define LV_DECODER_IMG_SIZE_EXPAND CONFIG_LV_DECODER_IMG_SIZE_EXPAND
+    #else
+        #define LV_DECODER_IMG_SIZE_EXPAND 0
     #endif
 #endif
 

